modules/internal/ChapelArray.chpl

```diff
@@ -2361,37 +2361,179 @@ module ChapelArray {
+      if isRectangularArr(this) {
+        if !value.dsiBoundsCheck(indices) {
+          if rank == 1 {
+            halt("array index out of bounds\n",
+                 "note: index was ", indices(1),
+                 " but array bounds are ", value.dom.dsiDim(1));
+          } else {
+            var istr = "";
+            var bstr = "";
+            for param i in 1..rank {
+              if i != 1 {
+                istr += ", ";
+                bstr += ", ";
+              }
+              istr += indices(i):string;
+              bstr += value.dom.dsiDim(i):string;
+            }
+            var dimstr = "";
+            for param i in 1..rank {
+              if !value.dom.dsiDim(i).boundsCheck(indices(i)) {
+                if dimstr == "" {
+                  dimstr = "out of bounds in dimension " + i:string +
+                           " because index " + indices(i):string +
+                           " is not in " + value.dom.dsiDim(i):string;
+                }
+              }
+            }
+            halt("array index out of bounds\n",
+                 "note: index was (", istr, ") ",
+                 "but array bounds are (", bstr, ")\n",
+                 "note: ", dimstr);
+          }
+        }
+      }
+    }
+
+    pragma "insert line file info"
+    pragma "always propagate line file info"
+    pragma "no doc"
+    proc checkSlice(d: domain, value) {
+      if isRectangularArr(this) {
+        var ok = true;
+        for param i in 1..rank {
+          ok &&= value.dom.dsiDim(i).boundsCheck(d.dsiDim(i));
+        }
+        if ok == false {
+          if rank == 1 {
+            halt("array slice out of bounds\n",
+                 "note: slice index was ", d.dsiDim(1),
+                 " but array bounds are ", value.dom.dsiDim(1));
+          } else {
+            var istr = "";
+            var bstr = "";
+            for param i in 1..rank {
+              if i != 1 {
+                istr += ", ";
+                bstr += ", ";
+              }
+              istr += d.dsiDim(i):string;
+              bstr += value.dom.dsiDim(i):string;
+            }
+            var dimstr = "";
+            for param i in 1..rank {
+              if !value.dom.dsiDim(i).boundsCheck(d.dsiDim(i)) {
+                if dimstr == "" {
+                  dimstr = "out of bounds in dimension " + i:string +
+                           " because slice index " + d.dsiDim(i):string +
+                           " is not in " + value.dom.dsiDim(i):string;
+                }
+              }
+            }
+            halt("array slice out of bounds\n",
+                 "note: slice index was (", istr, ") ",
+                 "but array bounds are (", bstr, ")\n",
+                 "note: ", dimstr);
+          }
+        }
+      }
+    }
+
+    pragma "insert line file info"
+    pragma "always propagate line file info"
+    pragma "no doc"
+    proc checkSlice(ranges...rank, value) where chpl__isTupleOfRanges(ranges) {
```

Comment:
Seems like we could re-use this code for the domain case. Could you call ``d.dsiDims()`` or create the tuple yourself, then pass it to this method?

Reply:
I agree and actually I had a version that worked that way originally. I had tried using `d.getIndices()` but that did not function correctly for sparse domains. It didn't occur to me to use `d.dsiDims()` though - which seems to work - so thanks for the tip!